### PR TITLE
Add actor that checks for obsolete .NET versions.

### PR DESCRIPTION
**Pull request by tmds**

### CODE REVIEW

**Comment by tmds** [comment]
Fixes https://issues.redhat.com/browse/RHELPLAN-92322.

@mreznik @mkluson ptal.

I don't speak python, and I'm not familiar with leapp. This is something to start from. I need your help to complete it.

cc @omajid

File: repos/system_upgrade/el8toel9/actors/dotnet/actor.py
Comment (omajid) [inline repos/system_upgrade/el8toel9/actors/dotnet/actor.py:6]: Perhaps we should extend this to include 2.1, 3.0, 5.0 as well?

**Comment by omajid** [comment]
Anyone know how we can test this? We would like to test this change manually to make sure the user experience matches what we expect. We are also interested in adding automated tests.

**Comment by MichalHe** [comment]
@omajid Hi,
in case you would like to manually test your PR I suggest using our internal vagrant box providing easy-to-setup testing environment. The repository with our vagrant box can be found the internal gitlab. Once you have our box up and running its a matter of of installing the RPM which gets automatically built after every push into the source branch of this PR. The RPM builds created from PRs following a specific naming scheme allowing them to be installed as follows: `dnf install -y leapp-upgrade-*pr867*` However, as the submitted code contains syntax errors, no builds for rhel8 are being built at the moment. Fixing them and (force) pushing the changes into the source branch will re-trigger builds, and the results should be available in the ` rpm-build:epel-8-x86_64` check found at the bottom of this PR. The link also contains instruction on how to install precisely the last build of this PR, without relying on wildcards.

Having installed the obsolete .NET version inside the running vagrant box + having installed a leapp-upgrade build from this PR should suffice you to manually test your code. The testing is then just a matter of running `leapp preupgrade` and inspecting whether the desired entry is created in the report. 

As per automatic testing - afaik all actors are covered with unit tests (examples can be found anywhere in the repository) and it is pretty much required for any new code before its get merged into the master. It is also possible to create and deliver integration tests using our tmt based framework that can be found in the internal gitlab.

I'll email to you the links to our rhel8 vagrant box and the integration testing framework as I would like to avoid sharing internal links.

**Comment by tmds** [comment]
I got the vm up and running.

When I run `leapp preupgrade`, the `dotnet_unsupported_versions_check` shows up:
```
...
====> * dotnet_unsupported_versions_check
        Check for installed .NET versions that are no longer supported.
...
```
However, it fails to detect the unsupported dotnet versions I have installed, and doesn't create a report for them.
```
# dnf list installed dotnet-runtime-*
Updating Subscription Management repositories.
Unable to read consumer identity

This system is not registered with an entitlement server. You can use subscription-manager to register.

Repository copr:copr.fedorainfracloud.org:group_oamg:leapp is listed more than once in the configuration
Installed Packages
dotnet-runtime-2.1.x86_64                                           2.1.30-1.el8_4                                           @rhel-8-for-x86_64-appstream-rpms
dotnet-runtime-3.0.x86_64                                           3.0.3-1.el8_1                                            @rhel-8-for-x86_64-appstream-rpms
dotnet-runtime-3.1.x86_64                                           3.1.22-4.el8                                             @rhel-8-for-x86_64-appstream-rpms
dotnet-runtime-5.0.x86_64                                           5.0.13-2.el8                                             @rhel-8-for-x86_64-appstream-rpms
dotnet-runtime-6.0.x86_64                                           6.0.2-2.el8                                              @rhel-8-for-x86_64-appstream-rpms
```

@MichalHe maybe if you look at the code, you can spot the issue?

Is there some documentation that describes how I can debug this locally? I'd like to build my changes in the VM and run `leapp preupgrade` without having to go through GitHub.

**Comment by pirat89** [comment]
@tmds you can check the `srpm` target in the Makefile. There should not be problem to modify the command to build package localy on RHEL system when you have installed all deps.

File: repos/system_upgrade/el8toel9/actors/dotnet/actor.py
Comment (pirat89) [inline repos/system_upgrade/el8toel9/actors/dotnet/actor.py:23]: the `InstaledRPM` cannot be used when the model is not in the tuple of consumed models (above, the `consumes` attribute is not even specified, which should be always ideally, even when empty)

File: repos/system_upgrade/el8toel9/actors/dotnet/actor.py
Comment (pirat89) [inline repos/system_upgrade/el8toel9/actors/dotnet/actor.py:15]:
```suggestion
    consumes = (InstalledRPM,)
    produces = (Report,)
```

File: repos/system_upgrade/el8toel9/actors/dotnet/actor.py
Comment (pirat89) [inline repos/system_upgrade/el8toel9/actors/dotnet/actor.py:28]: the msg should be a little bit more descriptive.
```suggestion
                reporting.Title('Unsupported .NET versions installed on the system.'),
```

File: repos/system_upgrade/el8toel9/actors/dotnet/actor.py
Comment (fernflower) [reply repos/system_upgrade/el8toel9/actors/dotnet/actor.py:15]: @tmds This should fix your issue with report not being created.
Long story short - unless you specify that your actor consumes InstalledRPM, has_package(InstalledRPM, 'smth') check won't work for you.

**Comment by pirat89** [comment]
@tmds one another question, is it expected to have this check for packages that are not provided by RH?

File: repos/system_upgrade/el8toel9/actors/dotnet/actor.py
Comment (tmds) [reply repos/system_upgrade/el8toel9/actors/dotnet/actor.py:15]: Thanks for fixing, and explaining the issue!

**Comment by tmds** [comment]
> is it expected to have this check for packages that are not provided by RH?

This actor checks for packages that are provided by RH. @omajid and I are part of the team that builds them.

**Comment by tmds** [comment]
Great! The actor detected the unsupported versions of .NET:

```
Risk Factor: high
Title: Unsupported .NET versions installed on the system.
Summary: The following versions of .NET are no longer supported :
    - 2.1
    - 3.0
    - 3.1
    - 5.0
Applications that use these runtimes will no longer work
and must be updated to target a newer version of .NET.
Key: 1d1c239ecef542adf0d69a6f2d4d9057b861b3ef
----------------------------------------
```

File: repos/system_upgrade/el8toel9/actors/dotnet/actor.py
Comment (fernflower) [reply repos/system_upgrade/el8toel9/actors/dotnet/actor.py:17]: Yes, the tags are correct.

**Comment by pirat89** [comment]
> > is it expected to have this check for packages that are not provided by RH?
> 
> This actor checks for packages that are provided by RH. @omajid and I are part of the team that builds them.

That's why I am asking because the current behaviour will raise the report even for packages of the same name provided by other vendors (or custom builds..). Regarding what you wrote, I expect you want to replace `InstalledRPM` by `InstalledRedHatSignedRPM` as only about these packages we can say they are provided by RH.

File: repos/system_upgrade/el8toel9/actors/dotnet/actor.py
Comment (tmds) [reply repos/system_upgrade/el8toel9/actors/dotnet/actor.py:36]: I'm going to leave it as is. I don't want to overwhelm the user with a list of packages, and it would be challenging to build the full list as dotnet has many packages for a single version.

File: repos/system_upgrade/el8toel9/actors/dotnet/tests/test_dotnet.py
Comment (fernflower) [inline repos/system_upgrade/el8toel9/actors/dotnet/tests/test_dotnet.py:34]: s/installed_versions/unsupported_versions?

File: repos/system_upgrade/el8toel9/actors/dotnet/tests/test_dotnet.py
Comment (fernflower) [inline repos/system_upgrade/el8toel9/actors/dotnet/tests/test_dotnet.py:39]: Sorry for nitpicking, could you please use one type of quotes for strings in the module? Either " or '?

**Comment by tmds** [comment]
> Regarding what you wrote, I expect you want to replace InstalledRPM by InstalledRedHatSignedRPM as only about these packages we can say they are provided by RH.

Understood. I'll switch to `InstalledRedHatSignedRPM`.

**Comment by fernflower** [comment]
@tmds Sorry for inconvenience with linter failures, they are unrelated to your patch. This PR should fix them, so if you would like to make a real run right away (meaning before PR869 is merged) you can cherry-pick it https://github.com/oamg/leapp-repository/pull/869

Once PR869 is merged (hopefully tomorrow) all that will be needed to make linting issues go away is rebase against master branch.

File: repos/system_upgrade/el8toel9/actors/dotnet/tests/test_dotnet.py
Comment (vinzenz) [reply repos/system_upgrade/el8toel9/actors/dotnet/tests/test_dotnet.py]: Actually we use single quotes for everything but doc strings, there might be some exceptions for some existing code that is used as a library, but we do usually insist on one style of quotes in the repository.

File: repos/system_upgrade/el8toel9/actors/dotnet/tests/test_dotnet.py
Comment (tmds) [reply repos/system_upgrade/el8toel9/actors/dotnet/tests/test_dotnet.py]: I tried using the `current_actor_context`.
Is that the way to go?

I need one of the maintainers to approve running workflows so I can see the CI results.

**Comment by tmds** [comment]
I don't see issues in the log that are related to the dotnet actor. I'm going to wait for the linting PR to get merged and then rebase.

**Comment by fernflower** [comment]
> I don't see issues in the log that are related to the dotnet actor. I'm going to wait for the linting PR to get merged and then rebase

You can always run tests locally without linter as `make test_no_lint`. Just make sure you have run `make install test-deps-fedora` before that.

**Comment by fernflower** [comment]
@tmds You can rebase now, the linting PR is in.

**Comment by tmds** [comment]
@fernflower can you approve the workflow run?

File: repos/system_upgrade/el8toel9/actors/dotnet/tests/test_dotnet.py
Comment (tmds) [inline repos/system_upgrade/el8toel9/actors/dotnet/tests/test_dotnet.py:4]: I don't understand this error:
```
F401 'leapp.snactor.fixture.current_actor_context' imported but unused
```
As far as I can tell, I'm using it in a similar way as other tests.

**Comment by tmds** [comment]
CI is green! :tada:

**Comment by tmds** [comment]
Is this ready to be merged?

File: repos/system_upgrade/el8toel9/actors/dotnet/actor.py
Comment (fernflower) [inline repos/system_upgrade/el8toel9/actors/dotnet/actor.py:40]: You could add troublesome packages to the report as well as RelatedResource('package', a_package).

**Review by fernflower — COMMENTED** [review]
Lgtm, just 1 small improvement suggestion.
Please also squash commits.

File: repos/system_upgrade/el8toel9/actors/dotnet/actor.py
Comment (tmds) [reply repos/system_upgrade/el8toel9/actors/dotnet/actor.py:40]: We're checking for the runtime packages, but there are other dependent packages which is why I did not add this.

For example, for 3.1:
```
dotnet-apphost-pack-3.1.x86_64                         3.1.18-1.fc33                       @updates                                                       
dotnet-hostfxr-3.1.x86_64                              3.1.18-1.fc33                       @updates                                                       
dotnet-runtime-3.1.x86_64                              3.1.18-1.fc33                       @updates                                                       
dotnet-sdk-3.1.x86_64                                  3.1.118-1.fc33                      @updates                                                       
dotnet-targeting-pack-3.1.x86_64                       3.1.18-1.fc33                       @updates                                                       
dotnet-templates-3.1.x86_64                            3.1.118-1.fc33                      @updates                                                       
```

I think adding all packages for all versions may be noise to the user. And only adding the runtime packages is too restrictive.

Does that make sense?

Comment (fernflower) [reply repos/system_upgrade/el8toel9/actors/dotnet/actor.py:40]: Sure, it makes sense. Thanks for explaining, I am ok with keeping it as is.

**Comment by tmds** [comment]
I've squashed the commits.

**Comment by fernflower** [comment]
@tmds Could you also add some description in the commit message that will explain what this patch is about? Something that would make it clear what's the actor's behaviour and the nature of reported problem (non-inhibitor, the upgrade continues even if troubesome packages are found) without looking at the actual code?

**Review by fernflower — APPROVED** [review]
lgtm

**Review by pirat89 — APPROVED** [review]

**Pull request merged**
